In [1]:
# Longest Increasing Subsequence
def LIS(nums):
    dp = [1] * len(nums)
    for i in range(len(nums)):
        for j in range(i):
            if nums[i] > nums[j]:
                dp[i] = max(dp[i], dp[j] + 1)
    return max(dp)

In [2]:
print(LIS([10, 9, 2, 5, 3, 7, 101, 18]))

4


The Longest Increasing Subsequence (LIS) finds the longest increasing path in the given list. Based on the given list, there are multiple valid LIS, with the longest possible paths having a length of 4.

The time complexity is exponential, O(n²), as it uses brute force by trying all subsequences. As for space complexity, it is O(n) due to the call stack.

This is relevant to my research by using the input as a sequence of robot sensor readings. LIS helps identify the longest sequence of trust or anomaly scores, which is useful for pattern detection in the state logs.

In [3]:
#Knapsack 0/1
def knapsack(W, wt, val, n):
    dp = [[0 for x in range(W + 1)] for x in range(n + 1)]
    for i in range(n + 1):
        for w in range(W + 1):
            if i == 0 or w == 0:
                dp[i][w] = 0
            elif wt[i - 1] <= w:
                dp[i][w] = max(val[i - 1] + dp[i - 1][w - wt[i - 1]], dp[i - 1][w])
            else:
                dp[i][w] = dp[i - 1][w]
    return dp[n][W]

In [4]:
print(knapsack(50, [10, 20, 30], [60, 100, 120], 3))

220


The 0/1 Knapsack problem maximizes the total value without exceeding the weight. In the example, the max value is 220 with the maximum weight being 50.

The time complexity is O(nW), appearing polynomial, and the space complexity is O(nW) as the memory usage grows with the input size.

This is relevant to my research as it models limited robot battery or bandwidth, choosing which alerts or tasks are most valuable to handle given resource limits. In a hospital rescue scene, it focuses on which alerts or tasks a drone should respond to first when energy is low.

In [5]:
# Edit Distance

def edit_distance(str1, str2):
    m, n = len(str1), len(str2)
    dp = [[0] * (n + 1) for _ in range(m + 1)]

    for i in range(m + 1):
        for j in range(n + 1):
            if i == 0:
                dp[i][j] = j
            elif j == 0:
                dp[i][j] = i
            elif str1[i - 1] == str2[j - 1]:
                dp[i][j] = dp[i - 1][j - 1]
            else:
                dp[i][j] = 1 + min(
                    dp[i][j - 1],
                    dp[i - 1][j],
                    dp[i - 1][j - 1]
                )

    return dp[m][n]

In [6]:
print(edit_distance("kitten", "sitting"))


3


The edit distance focuses on finding the minimum number of edits like insert, delete, and replace that are needed to convert one string to another.

The time complexity is O(m * n), which is quadratic based on the combined size of the inputs, and the space complexity is O(m * n), meaning the memory usage grows quadratically with respect to the two input parameters.

This is relevant to my research as it is useful for comparing robot plans or sensor logs over time. It can help detect anomalies between expected and observed behaviors, the edit distance can signal a deviation.